### DISCUSSION TITLE
Different Files are Identical?

### DISCUSSION THREAD

**Discussion by Hammerhead10002** [WinMerge/winmerge#768] [Q&A]
Hello All,

I really like WinMerge and thank everyone for their efforts. I have a question please.

Winmerge Compare tells me that "Text Files are Different" (note that attributes appear to be the same).
![WM_04](https://user-images.githubusercontent.com/83463874/116723297-5a0a6880-a9ad-11eb-9d2b-e829b3f0486c.JPG)

When I double click to open them, Winmerge tells me that "The Selected Files are Identical".
![WM_05b](https://user-images.githubusercontent.com/83463874/116723829-e3219f80-a9ad-11eb-955c-851bffec2986.jpg)

Any ideas? If it makes any difference, I believe these files were saved from an Android Phone. 

Thank you in advance for your consideration. 
(I am using v2.18.8.0.. will update today).

.

**Comment by sdottaka** [comment]
Would you please show me the result of comparing the bookmark.txt file as a binary as below?

![image](https://user-images.githubusercontent.com/98126/116766106-ac0dc580-aa63-11eb-977a-c8cf273fd20d.png)

**Reply by Hammerhead10002** [reply]
Hello sdottaka - thanks a lot for the reply! Sorry for the length....

This became a bit erratic:
- I updated from v2.16.8 (correction to my above mistake) to the current stable v2.16.12 last night.
- I then ran the same project overnight (as above). It is worth mentioning that it is big- about 3TB total; 16 hrs of processing. (I read the Release Notes regarding crashes with large files). It did not crash.... but.....
- Two of the three offending files shown above were now listed as identical. Only one different file in a deeply nested directory.
- I ran your suggested "Compare As Binary" as a separate project- the remaining offending file came up as "identical".
- I went back to the view of the original project, and oddly, the offending upper parent directories were still red, the immediate parent directory was now green, but the offending file was nowhere to be found? Did WinMerge hide it because the "Compare As" came up as identical? Why do the upper level directories remain red, while lower level turned green (and display "different")?
- I the ran a new project, directing the Compare to include only the parent directories of the 3 original troubled files (they are all in the same nested directory). It came up clean with NO different files?
- Changing view parameters oddly changed results back-and-forth from different to identical (when I could see the disappearing file).
A few snips:
After last night's run, 2 of the original 3 "different" files are now identical; only 1 remains:
![20210501_WM_Snip_01e](https://user-images.githubusercontent.com/83463874/116790072-6e6b6580-aa80-11eb-8d50-d07bc206ba80.jpg)

Ran a separate project to "Compare As Binary" as you suggested:
![20210501_WM_Snip_02e](https://user-images.githubusercontent.com/83463874/116790694-b63fbc00-aa83-11eb-938a-1b4670809c66.jpg)

"Different" File changed from red to green "Identical":
![20210501_WM_Snip_03e](https://user-images.githubusercontent.com/83463874/116790735-e6875a80-aa83-11eb-9bdd-0947cd6634bc.jpg)

Immediate Parent directories are now green (but higher level remain red), file no longer visible:
![20210501_WM_Snip_04e](https://user-images.githubusercontent.com/83463874/116790790-35cd8b00-aa84-11eb-97e3-f93f3427938a.jpg)

Maybe this behavior is normal, but it seems erratic to me?
- After the upgrade, the new full overnight Project only came up with one of the three original different files.
- A secondary Project to compare a single top level directory containing the troubled files came up with NO differences.
- Switching repeatedly between Tree Mode on/off and "Show identical Items" on/off  ("Show Different Items" always on) seemed to cause erratic displays... sometimes the same, sometimes not. Maybe the Project is just too big?
- I'm thinking of running a hash on the important files to verify, even though Teracopy supposedly did that.
- I had originally stated these files possible came from an Android phone... but I guess probably they are from a PC (long story).

Your thoughts would be appreciated! Is it me? Is there a bug? The project is too large?

**Comment by sdottaka** [comment]
Probably the following fixes have fixed this issue.

https://github.com/WinMerge/winmerge/commit/0270fd762ade2c6340db0e3d6126effdfb0ba58c

All of the files shown in the image above are likely to be UTF-16 files with only a BOM and no content.
There was a problem that the comparison result was not stable when comparing such files.

The latest unreleased version installer is here:
https://ci.appveyor.com/project/sdottaka/winmerge/build/artifacts

**Comment by bnbarnes** [comment]
I am seeing this same issue up through 2.16.18.0
All comparisons ignore nothing and even binary comparison show identical but yellow highlights show the differences
![image](https://user-images.githubusercontent.com/4754975/154747954-71343e79-cffe-48ad-b18e-bdbdd93e002f.png)

**Reply by sdottaka** [reply]
@bnbarnes Are the Line filters, Substitution filters, and prediffer plug-ins enabled? They may be causing the differences to be ignored.
![image](https://user-images.githubusercontent.com/98126/154776133-909d5161-9e94-4003-974e-f1c56392a368.png)
![image](https://user-images.githubusercontent.com/98126/154776177-bea0bd84-d257-4896-94f5-c9006f5dd990.png)
![image](https://user-images.githubusercontent.com/98126/154776223-0c33b6f1-aa5c-415b-896d-0d00c9a92530.png)

**Comment by bnbarnes** [comment]
I have moved on from this issue and resolved using meld. I cannot reproduce right now.
These are the only filter settings I had:
![image](https://user-images.githubusercontent.com/4754975/154777802-8b76a8a9-2985-4f14-a85b-865a04cce353.png)
![image](https://user-images.githubusercontent.com/4754975/154777827-2dc5c68e-5882-44bc-91ae-560d05c57772.png)
![image](https://user-images.githubusercontent.com/4754975/154777858-f8577d06-f102-4d2e-b3f8-455f8cdc529f.png)
![image](https://user-images.githubusercontent.com/4754975/154777880-78c0e2e8-7165-476c-bd65-a95539ee089a.png)
Regex was the only filter on but it can't account the for alphas in  -x86_64 vs -i386

**Comment by sdottaka** [comment]
The regular expression `[0-9]. *` matches `-7-x86_64 ...` and `-7-i386 ...` as shown below, so the difference is ignored. 

![image](https://user-images.githubusercontent.com/98126/154779245-f82e39f4-d86d-49a5-8f94-34854239c8ff.png)
